central/sensor/service/connection/initsync_manager.go

```diff
@@ -0,0 +1,61 @@
+package connection
+
+import (
+	"fmt"
+
+	"github.com/stackrox/rox/pkg/env"
+	"github.com/stackrox/rox/pkg/set"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+type initSyncManager struct {
+	mutex sync.Mutex
+
+	maxSensors int
+	sensors    set.StringSet
+}
+
+// NewInitSyncManager creates an initSyncManager with max sensors
+// retrieved from env variable, ensuring it is non-negative.
+func NewInitSyncManager() *initSyncManager {
+	maxSensors := env.CentralMaxInitSyncSensors.IntegerSetting()
+	if maxSensors < 0 {
+		panic(fmt.Sprintf("Negative number is not allowed for max init sync sensors. Check env variable: %q", env.CentralMaxInitSyncSensors.EnvVar()))
+	}
+
+	return &initSyncManager{
+		maxSensors: maxSensors,
+		sensors:    set.NewStringSet(),
+	}
+}
+
+func (m *initSyncManager) Add(clusterID string) bool {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if m.maxSensors == 0 {
+		return true
+	}
+
+	if m.sensors.Contains(clusterID) {
```

Comment:
when would this happen 🤔

Reply:
My concern was when a connection fails during an init sync:
- a cluster (sensors) is already in the set (I'm not sure if we are calling CloseConnection on a failed connection?)
- the sensor will try to re-connect
- since check for limit is before adding cluserID to set
=> we could end up rejecting a sensor that is already in the set

I have looked at the code a bit more. If I'm not wrong `conn.Run` is blocking and I can add remove after it. That should clean up sensors if a connection fails during init sync.

https://github.com/stackrox/stackrox/blob/743dea7ad2b63435ea6c54ba95157026e05d714b/central/sensor/service/connection/manager_impl.go#L275-L277